### ISSUE TITLE
Find missing visits in Run2.2i Y02 dataset

### ISSUE THREAD

**Issue by jchiang87** [LSSTDESC/DC2-production#387]
Here are the depth maps for the Y02 data processed for Run2.2i as of today:
![Run2 2i_y02_log10_depths](https://user-images.githubusercontent.com/1994473/76535987-f768ad00-6438-11ea-8349-1c27f5626939.png)
We have ascertained that there are ~550 visits in the minion_1016 ops db that we had intended to process but which do not appear in the Run2.2i processed calexps.

The missing visits seem to be the ones we generated for Run2.1.1i, which we omitted from Run2.2i to avoid redoing the image sims.   We'll use this issue to document what files are missing, and what's available from Run2.1.1i, both in terms of raw files and in processed data.

**Comment by johannct** [comment]
relevant pointer from the record of transfers NERSC <-> IN2P3 : https://github.com/LSSTDESC/ImageProcessingPipelines/issues/86#issuecomment-521841925

**Comment by johannct** [comment]
this resulted in 257 visits processed, the list being
390560
398414
398415
400423
400424
400425
400439
400459
400460
400461
401608
401609
401614
401615
401650
401651
401661
406954
406977
406993
406994
407915
407918
407927
407928
407941
407942
407945
407946
410959
410960
410985
410986
410995
414861
414868
414869
414892
414908
414909
414910
414912
417004
417005
417052
417053
418192
418203
418204
418966
418999
419025
419026
419027
419063
419068
419774
419908
420801
420802
420828
420879
421642
421646
421647
421676
421677
421680
421684
421685
421689
421690
425506
425510
425525
425529
426670
426671
426924
426925
426926
426971
426972
426973
427029
427643
427672
431363
431364
431402
431404
433045
433073
433075
433952
433953
433954
433955
433956
433984
433985
433986
433990
436536
436537
436540
436541
436580
436581
436584
436585
440972
443017
443125
443126
443157
443158
443159
444043
444063
444079
444080
444111
444112
444707
444713
444714
444715
444716
447511
447521
447522
451658
451696
451697
451698
451755
451797
452558
452594
453605
456674
456675
456683
456684
456708
456717
456743
456744
456745
456746
456770
456771
456773
456774
456845
457675
457682
457683
457710
457711
457712
457715
457716
458289
458297
458298
458309
458311
458564
458603
458604
458647
460322
462664
462666
462669
462670
462705
462706
462708
462713
462715
462716
462717
466276
466277
466320
466321
466323
466717
466747
466748
466750
466757
466759
467700
468251
468252
468293
468294
471902
471929
471930
471931
471934
471935
471957
471958
471961
471962
471999
472134
472136
472137
472180
472181
474334
474335
475690
476317
476321
478930
478990
478991
479027
479028
479266
479298
479512
479542
479662
479665
479696
480300
480301
480302
480303
480341
480344
480679
484233
484234
484235
484237
484263
484264
484265
484267
484277
491590
492034
492035
496961
496988
497008
497043

**Comment by jchiang87** [comment]
Here's a file with the missing y02 visits.  The columns are visit number and MJD.
[Run2.2i_y02_missing_visits.txt](https://github.com/LSSTDESC/DC2-production/files/4324989/Run2.2i_y02_missing_visits.txt)

Here is the code that generated this list:
```
import os
import glob
import sqlite3
import pandas as pd

# Find the visits that were processed
visit_dirs = glob.glob('/global/cfs/cdirs/lsst/production/DC2_ImSim/Run2.2i/desc_dm_drp/v19.0.0-v1/rerun/run2.2i-calexp-v1/calexp/*')
processed_visits = set(int(os.path.basename(_).split('-')[0]) for _ in visit_dirs)

# Extract the y02 visits from the minion_1016 db
with sqlite3.connect('/global/homes/j/jchiang8/scratch/desc/Run2.2i/minion_1016_desc_dithered_v4_trimmed.db') as conn:
    df = pd.read_sql('select * from summary', conn)
t0 = min(df['expMJD'])

my_df = df.query(f'{t0 + 365} < expMJD < {t0 + 2*365}')
minion_visits = set(my_df['obsHistID'])

# Find the missing visits
missing_visits = sorted(list(minion_visits.difference(processed_visits)))

# Write a file with the visits and MJDs
expMJDs = dict(zip(my_df['obsHistID'], my_df['expMJD']))

missing_visits = {visit: expMJDs[visit] for visit in missing_visits}
with open('Run2.2i_y02_missing_visits.txt', 'w') as output:
    for visit, expMJD in missing_visits.items():
        output.write(f'{visit}   {expMJD:.2f}\n')
```

**Comment by johannct** [comment]
ok my list of visits is entirely inside yours, Jim. So we need to make sure we understand what is going on with the other ones : the ones downloaded to CC go from 00385844 to 00445379 according to the naming convention in the paths at CC.

**Comment by villarrealas** [comment]
[all_visits_to_date_031220.txt](https://github.com/LSSTDESC/DC2-production/files/4325316/all_visits_to_date_031220.txt)
At the request of Jim, here is a file containing all visits simulated to date!

**Comment by jchiang87** [comment]
Based on Antonio's list of simulated visits and on the [master list of minion_1016 visits](https://github.com/LSSTDESC/sims_GCRCatSimInterface/blob/master/workspace/run2.1/data/master_obshistid_list.txt), there are 540 visits missing in total through year 3.  However, about half of those visits (specifically all of the y01 and y03 missing ones) are on the periphery of the DC2 300 sq degree region, and 256 would need to be simulated to fill the hole in y02.   Here is the summary plot showing the distribution of missing visits as function of MJD, and the centers of the visits (i.e., the pointing directions) for each year.
![missing_visits_summary](https://user-images.githubusercontent.com/1994473/76549609-fd1cbd80-644d-11ea-8ed8-6c940b9c4d28.png)

STOP PRESS:  Eve recalled that we found a problem with the redshifts in the instance catalogs used for Run2.1.1i (`redshift_true` was used instead of `redshift`, which includes peculiar velocities), so we would need to re-simulate all of the visits in hole, nominally the 555 I posted above.

**Comment by villarrealas** [comment]
Heather and I are pulling Y2 Run2.2 instance catalogs to tape to prep for any necessary resimulation and verify that these visits did recieve these updated instance catalogs. Now that the memory is flowing, I thought we had chosen to resim y2 from scratch because of this issue and I am unsure why we are missing these visits...

**Comment by jchiang87** [comment]
Here's a plot of the pointing directions of the 555 y02 "missing" visits:
![y02_missing_visits](https://user-images.githubusercontent.com/1994473/76563505-c900c700-6464-11ea-9b52-be388378bde8.png)
As noted above, the visits in this list are the ones in minion_1016 that are not in registry.sqlite3 for y02.   Assuming we will skip the visits outside of the DC2 boundary, there are 256 visits to simulate, which are plotted in red.   Here is the list:
[y02_missing_visits_to_simulate.txt](https://github.com/LSSTDESC/DC2-production/files/4326475/y02_missing_visits_to_simulate.txt)

**Comment by heather999** [comment]
The Y02 Run2.2i instance catalogs are being copied from HPSS and are appearing here: `/global/projecta/projectdirs/lsst/production/DC2_ImSim/Run2.2i/InstCat/copyFromHPSS/y02_191109`  I expect the transfer to finish in the next ~4 hours, and then it's just a matter of extracting the tarballs.

**Comment by heather999** [comment]
Antonio tells me that what I pulled from HPSS and put into `/global/projecta/projectdirs/lsst/production/DC2_ImSim/Run2.2i/InstCat/copyFromHPSS/y02_191109` - contains none of the missing visits.  This is a bit mystifying, because I copied these files directly from the instance catalog area Scott used to generate the instance catalogs: `/global/cscratch1/sd/desc/HPSS/dc2/run2.2i/instCat/y02`
However there is another stash of instance catalogs on projecta: `/global/projecta/projectdirs/lsst/production/DC2_ImSim/Run2.2i/InstCat/WFD/y02` which is owned by the `descim` account, that I have no knowledge of.  I don't know where this set of instance catalogs came from and why it seems to be completely distinct from what I thought was the full set.

A couple of things.. we need to verify that the instance catalogs contained in `/global/projecta/projectdirs/lsst/production/DC2_ImSim/Run2.2i/InstCat/WFD/y02` should be used to generate the missing visits - are they configured correctly? Who can check this?  And secondly - it would be nice to figure out where this stash came from.  I'll set out to add this set to HPSS so we have them all stored for posterity.
If there are no objections - I will remove the copy of y02 instance catalogs that I already pulled from HPSS to free up some disk space, as it seems we don't need them.

**Comment by katrinheitmann** [comment]
I hope we did use the new instance catalogs that Joanne generated and 
not the incorrect once from Scott. I also hope that the instance 
catalogs we used were not purged by now if they were not moved.

Antonio, can you clarify what you used?

**Comment by heather999** [comment]
I think what I stored to HPSS is what Joanne generated..  Though I would like to clarify what area was used for her instance catalog generation - I believe it was in the same area that Scott originally used.  These files were copied to tape on Jan 31, 2020 from `/global/cscratch1/sd/descim/instcat_y02_191109` Note the corrected path from what I commented above.

**Comment by jchiang87** [comment]
The fact that those file retrieved from HPSS do not have the missing visits explains why the images for those visits were not generated in the first place, since the intent for Run2.2i was to generate imsim data for **everything**.   I'd like to have a closer look at the unpacked files myself first, but I expect there is something in the instance catalog generation scripts that skipped the Run2.1.1i visits that was still enabled when Joanne remade the instance catalogs with the corrected redshifts, etc..

**Comment by villarrealas** [comment]
I can confirm that we used the instance catalogs that were then stored to tape. That gives us a good explanation as to why we missed these visits at least, since my worklist pulls directly from the list of instance catalogs and then trims based on the sensor region.

As to why we seemingly have the remaining instance catalogs in another folder... that I cannot answer. My only possible guess is that we separated them out, but I do not recall doing this and the folder naming is not consistent with this hypothesis.

**Comment by jchiang87** [comment]
> As to why we seemingly have the remaining instance catalogs in another folder

**Comment by jchiang87** [comment]
Here is the provenance of the instcats in `/global/projecta/projectdirs/lsst/production/Run2.2i/InstCat/WFD/y02`:

https://lsstc.slack.com/archives/C77DDKZHR/p1571690314032400

Scott wrote them and they were the ones used to find and diagnose the redshift problem (among other things).

**Comment by villarrealas** [comment]
Wait, is that perhaps why these were left out of the remaining instance catalogs? These were final  versions and it was just chosen not to redo the work during instance catalog generation?

**Comment by jchiang87** [comment]
That seems to be the case.

**Comment by jchiang87** [comment]
@JoanneBogart has agreed to generate new instance catalogs for the 513 visits that are needed using the list posted above.   Once those are available, we can use them to generate the images for those visits.

**Comment by villarrealas** [comment]
Sounds like a plan to me, though I thought these would be equivalent to those in `/global/projecta/projectdirs/lsst/production/Run2.2i/InstCat/WFD/y02`. Were additional changes needed from that baseline?

**Comment by jchiang87** [comment]
The files in that directory used incorrect redshifts as we found by comparing the implied fluxes with the truth catalog values, and there are only 279 of them.

**Comment by villarrealas** [comment]
Hrm. I would have sworn I remember looking at it and having 500+ instance catalogs. Still, if they were with incorrect redshifts, we can wait for Joanne to generate the remaining 513 visits worth.

**Comment by JoanneBogart** [comment]
The job is running now.  It should be done in a few hours.

**Comment by JoanneBogart** [comment]
500 catalogs were written without error.  The tar files are available at
`/global/cscratch/descim/y02_instCat_missing`

I'll redo the others and put them in a separate directory.

**Comment by katrinheitmann** [comment]
When you say "the others" you mean the other 13? Just to be sure, we only have to redo the "hole" now, everything else is correct, right? (It's sometimes hard to follow the discussions). Thanks!

**Comment by JoanneBogart** [comment]
Yes.   Just the other 13 from Jim's list above.

**Comment by JoanneBogart** [comment]
Those 13 are now done as well.  The tar files are in `/global/cscratch/descim/y02_instCat_missing_remaining`

**Comment by jchiang87** [comment]
With the instance catalogs generated and simulations running, I'm closing this issue.  If we need to track anything having to do with sims or image processing, we can open an new issue.
